### PR TITLE
Feature/dgraph upgrade

### PR DESCRIPTION
**Pull request by bendeeson**
**[Draft PR]**

Updates to opendialog core to support dgraph v20.11.0.

* Updated 'Graph' module to support dgraph json response format changes related to facets. See https://discuss.dgraph.io/t/facets-format-in-mutation-requests-and-query-responses/6416
* Updated dgraph-docker dependency to dgraph v20.11.0
* Updated dgraph client to send authorization header on each request. This allows it to perform alter operations. (E.g initSchema)

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "guzzlehttp/guzzle": "~6.3",
     "justinrainbow/json-schema": "^5.2",
     "spatie/laravel-activitylog": "^3.8",
-    "opendialogai/dgraph-docker": "1.2.5",
+    "opendialogai/dgraph-docker": "dev-feature/dgraph-v20.11.0",
```
Comment (bendeeson) [inline composer.json:13]: Reminder to swap this to a tag once `feature/dgraph-v20.11.0` is merged.

**Comment by istos** [comment]
Nice work Ben 👍

File: src/Graph/GraphServiceProvider.php [hunk 2]
```diff
@@ -25,11 +25,14 @@ public function register()
                 $schema .= config('opendialog.graph.custom_schema');
             }
 
-            return new DGraphClient(
-                config('opendialog.core.DGRAPH_URL'),
-                config('opendialog.core.DGRAPH_PORT'),
-                $schema
-            );
+            if ($dgraphAuthToken = config('opendialog.core.DGRAPH_AUTH_TOKEN')) {
```
Comment (bendeeson) [inline src/Graph/GraphServiceProvider.php:28]: The DGRAPH_AUTH_TOKEN config value (read from `DGRAPH_AUTH_TOKEN` env var) is required. An exception is thrown if it is not set.

File: src/Graph/DGraph/DGraphClient.php [hunk 3]
```diff
@@ -37,14 +37,19 @@ class DGraphClient
...
+     * @param  string  $dGraphPort
+     * @param  string  $dgraphAuthToken
+     * @param  string  $schema
      */
-    public function __construct(string $dgraphUrl, string $dGraphPort, string $schema)
+    public function __construct(string $dgraphUrl, string $dGraphPort, string $dgraphAuthToken, string $schema)
     {
         $client = new Client([
-            'base_uri' => $dgraphUrl . ":" . $dGraphPort
+            'base_uri' => $dgraphUrl . ":" . $dGraphPort,
```
Comment (bendeeson) [inline src/Graph/DGraph/DGraphClient.php:49]: Every DGraph request sends the `X-Dgraph-AuthToken` header, though this is strictly only neccessary for `admin` operations: (See https://dgraph.io/docs/deploy/dgraph-administration/#securing-alter-operations)

File: src/ContextEngine/Tests/UserServiceTest.php [hunk 4]
```diff
@@ -285,7 +285,7 @@ public function testFollowedByAndPrecededBy()
         ]));
 
         $this->assertArrayHasKey(Model::FOLLOWED_BY, $response->getData()[0]);
-        $this->assertArrayHasKey(ModelFacets::facet(Model::FOLLOWED_BY, ModelFacets::CREATED_AT), $response->getData()[0]);
+        $this->assertArrayHasKey(ModelFacets::facet(Model::FOLLOWED_BY, ModelFacets::CREATED_AT), $response->getData()[0][Model::FOLLOWED_BY]);
```
Comment (bendeeson) [inline src/ContextEngine/Tests/UserServiceTest.php:288]: The shape of facet responses changed in intermediate dgraph versions. See https://github.com/dgraph-io/dgraph/pull/5424

File: composer.json (hunk 1 above)
Comment (stu-greenshoots) [reply composer.json:13]: @bendeeson - this is tagged now, so you can update with the `20.11.0` tag

File: Tests/Feature/ConfigTest.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        return [
+            CoreServiceProvider::class,
+            GraphServiceProvider::class,
+        ];
+    }
+
+    /**
+     * Check that absence of DGRAPH_AUTH_TOKEN throws an exception.
+     */
+    public function testMissingDGraphAuthTokenConfig() {
```
Comment (stu-greenshoots) [inline Tests/Feature/ConfigTest.php:27]: Nice to see this test, thanks!

File: .circleci/config.yml [hunk 6]
```diff
@@ -57,7 +57,7 @@ jobs:
       - run:
           name: "Setup custom environment variables"
           command: |
-            echo 'export OVERWRITE_DGRAPH_URL="http://localhost"' >> $BASH_ENV; echo 'export OVERWRITE_DGRAPH_PORT="8080"' >> $BASH_ENV
+            echo 'export OVERWRITE_DGRAPH_URL="http://localhost"' >> $BASH_ENV; echo 'export OVERWRITE_DGRAPH_PORT="8080"' >> $BASH_ENV; echo 'export OVERWRITE_DGRAPH_AUTH_TOKEN=circlecidgraphauthtoken' >> $BASH_ENV
```
Comment (stu-greenshoots) [inline .circleci/config.yml:60]: Glad this works 👍

**Review by stu-greenshoots — APPROVED** [review]
Great Ben, looks good to me.

Once you have updated the composer dependency to use a tag, I'm happy for you to merge 👍

**Pull request merged**